File: crates/marmot-uniffi/tests/options_smoke.kt
package dev.ipf.marmotkit

import java.nio.file.Files

private class OptionsStore : SecretStore {
    override fun hasSecretForLabel(label: String): Boolean = label == "probe"
    override fun hasSecretForAccountId(accountIdHex: String): Boolean = false
    override fun writeSecret(label: String, accountIdHex: String, secretKeyHex: String) {}
    override fun loadSecret(label: String, accountIdHex: String): String = "callback-probe"
    override fun removeSecret(label: String, accountIdHex: String) {}
}

fun main() {
    val defaults = MarmotOptions()
    check(defaults.relayPolicy == null && defaults.cursorPersistence == null)
    check(defaults.clientName == null && defaults.secretStore == null)
    val options = MarmotOptions(
        relayPolicy = RelayPolicyFfi.ALLOW_LOOPBACK_RELAYS_AND_BLOBS,
        cursorPersistence = CursorPersistenceFfi.FROZEN,
        clientName = "whitenoise",
    )
    val copy = FfiConverterTypeMarmotOptions.lift(FfiConverterTypeMarmotOptions.lower(options))
    check(copy.relayPolicy == options.relayPolicy && copy.cursorPersistence == options.cursorPersistence)
    check(copy.clientName == "whitenoise")
    // Foreign callbacks are lowered into Rust by the constructor, not locally lifted.
    copy.secretStore = OptionsStore()
    val root = Files.createTempDirectory("marmot-options-").toFile()
    try {
        Marmot.newWithConfiguration(root.path, listOf("ws://127.0.0.1:1"), copy).use { kit ->
            check(kit.listAccounts().isEmpty())
        }
    } finally {
        root.deleteRecursively()
    }
    println("Kotlin runtime options defaults and host-store construction passed")
}

File: crates/marmot-uniffi/tests/options_smoke.swift
import Foundation

private final class OptionsStore: SecretStore, @unchecked Sendable {
    func hasSecretForLabel(label: String) throws -> Bool { label == "probe" }
    func hasSecretForAccountId(accountIdHex: String) throws -> Bool { false }
    func writeSecret(label: String, accountIdHex: String, secretKeyHex: String) throws {}
    func loadSecret(label: String, accountIdHex: String) throws -> String { "callback-probe" }
    func removeSecret(label: String, accountIdHex: String) throws {}
}

@main
struct OptionsSmoke {
    static func main() throws {
        let defaults = MarmotOptions()
        precondition(defaults.relayPolicy == nil && defaults.cursorPersistence == nil)
        precondition(defaults.clientName == nil && defaults.secretStore == nil)
        let options = MarmotOptions(
            relayPolicy: .allowLoopbackRelaysAndBlobs,
            cursorPersistence: .frozen,
            clientName: "whitenoise"
        )
        var copy = try FfiConverterTypeMarmotOptions.lift(FfiConverterTypeMarmotOptions.lower(options))
        precondition(copy.relayPolicy == .allowLoopbackRelaysAndBlobs)
        precondition(copy.cursorPersistence == .frozen && copy.clientName == "whitenoise")
        // Foreign callback handles must cross into Rust; local lower/lift is
        // only valid here while the callback field is nil.
        copy.secretStore = OptionsStore()
        let root = FileManager.default.temporaryDirectory.appendingPathComponent(UUID().uuidString)
        defer { try? FileManager.default.removeItem(at: root) }
        let kit = try Marmot.newWithConfiguration(
            rootPath: root.path, relayUrls: ["ws://127.0.0.1:1"], options: copy
        )
        let accounts = try kit.listAccounts()
        precondition(accounts.isEmpty)
        print("Swift runtime options defaults and host-store construction passed")
    }
}
